onnxscript/optimizer/_constant_folding.py

```diff
@@ -0,0 +1,712 @@
+# partial evaluation and/or static program analysis (abstract interpretation).
+
+# A partial-evaluator function takes a node, a RewriterContext, OptimizerState and returns
+# a Replacement for the node or None (if no replacement is needed). It may also return just
+# the ir.Value or ir.Values to replace the output values of the node, when the new nodes
+# can be inferred from the RewriterContext used to build the new nodes.
+
+ReturnValue = Union[Replacement, Sequence[ir.Value], ir.Value, None]
+PartialEvaluatorFunction = Callable[
+    [ir.Node, orp.RewriterContext, OptimizerState], ReturnValue
+]
+
+
+@dataclasses.dataclass
+class PartialEvaluator:
+    """A class that represents a partial-evaluator for a particular op.
+
+    It is applicable for a specific version range (min_version, max_version) of the op.
+    The min_version and max_version can be None, indicating that there is no version
+    constraint in that direction.
+    """
+
+    min_version: int | None
+    max_version: int | None
+    function: PartialEvaluatorFunction
+
+    def valid_for(self, version: int) -> bool:
+        """Returns True if this evaluator is applicable for the given version."""
+        return (self.min_version is None or version >= self.min_version) and (
+            self.max_version is None or version <= self.max_version
+        )
+
+
+class PartialEvaluatorRegistry:
+    """A class that maintains a registry of evaluators for ops."""
+
+    def __init__(self):
+        self.op_evaluators: dict[tuple[str, str], list[PartialEvaluator]] = {}
+
+    def lookup_evaluators(self, domain: str, opname: str, version: int):
+        evaluator_list = self.op_evaluators.get((domain, opname), [])
+        return [
+            evaluator.function for evaluator in evaluator_list if evaluator.valid_for(version)
+        ]
+
+    def register(self, opname: str, domain: str = "", version=None):
+        if (domain, opname) not in self.op_evaluators:
+            evaluator_list = []
+            self.op_evaluators[(domain, opname)] = evaluator_list
+        else:
+            evaluator_list = self.op_evaluators[(domain, opname)]
+        if version is None:
+            min_version = None
+            max_version = None
+        elif isinstance(version, int):
+            min_version = version
+            max_version = version
+        elif isinstance(version, tuple):
+            min_version, max_version = version
+
+        def decorator(function: PartialEvaluatorFunction) -> PartialEvaluatorFunction:
+            evaluator_list.append(PartialEvaluator(min_version, max_version, function))
+            return function
+
+        return decorator
+
+
+registry: PartialEvaluatorRegistry = PartialEvaluatorRegistry()
+
+register = registry.register
+
+
+def _get_numpy_value(val: ir.Value) -> np.ndarray | None:
+    const_value = val.const_value
+    if hasattr(const_value, "numpy"):
+        return const_value.numpy()
```

Comment:
`val.const_value` should always be an `ir.TensorProto` type or None so it should always have `numpy()` when it is not None. Currently the type is not enforced for runtime efficiency, but we can add more type checking under the `DEBUG` mode